File: flowersec-go/internal/protocolv4/application_reference_test.go
package protocolv4

// Independent test-only application byte relations. Authentication, actual
// channel/serial ownership, admission, codec execution and publication remain external.
import (
	"bytes"
	"encoding/json"
	"slices"
	"strconv"
	"testing"
)

type applicationVariant struct {
	Code            uint64
	Fields          []uint64
	Constants       map[string]uint64
	Request         string
	SDKError        bool   `json:"sdk_error"`
	BusinessError   bool   `json:"application_error"`
	MaxEncodedBytes uint64 `json:"max_encoded_bytes"`
}
type applicationRegistry struct {
	Kinds  map[string]applicationVariant
	Policy struct {
		ExecutionRequests []string                     `json:"execution_requests"`
		ContractVariants  map[string]map[string]uint64 `json:"contract_variants"`
	}
	SDKErrors struct {
		PayloadSchema string `json:"payload_schema"`
	} `json:"sdk_errors"`
	SDKCodes map[string]uint64 `json:"sdk_error_codes"`
}
type applicationHeader struct {
	kind  string
	value *cborRefValue
	raw   []byte
}
type applicationReference struct {
	*domainReference
	application applicationRegistry
	names       map[string]map[string]uint64
}

func newApplicationReference(t testing.TB) *applicationReference {
	t.Helper()
	r := &applicationReference{domainReference: newDomainReference(t), names: map[string]map[string]uint64{}}
	if err := json.Unmarshal([]byte(ApplicationHeaderRegistryJSON), &r.application); err != nil {
		t.Fatal(err)
	}
	for name, descriptor := range r.registry.Maps {
		r.names[name] = map[string]uint64{}
		for id, field := range descriptor.Fields {
			n, err := strconv.ParseUint(id, 10, 64)
			if err != nil {
				t.Fatal(err)
			}
			r.names[name][field.Name] = n
		}
	}
	return r
}

// Callers use generated field names only after a complete map passes wireMap.
// Optional header fields are explicitly nil-checked at their use sites.
func (r *applicationReference) value(name string, v *cborRefValue, field string) *cborRefValue {
	id, ok := r.names[name][field]
	if !ok {
		panic("unresolved generated field")
	}
	return cborLookup(v, id)
}
func (r *applicationReference) bound(name string) uint64 {
	bound := r.registry.Maps[name].MaxEncodedBytes
	if bound == nil {
		panic("unresolved generated bound")
	}
	return *bound
}
func applicationCapture(input []byte, bound uint64) ([]byte, error) {
	if uint64(len(input)) > bound {
		return nil, cborRefError("application_input_size")
	}
	return bytes.Clone(input), nil
}
func (r *applicationReference) rawMap(name string, input []byte) (*cborRefValue, error) {
	raw, err := applicationCapture(input, r.bound(name))
	if err != nil {
		return nil, err
	}
	return r.wireMap(raw, name, cborShapeContext{}, r.bound(name))
}
func (r *applicationReference) hash(name string, args map[string]any) ([]byte, error) {
	result, err := r.evaluate(name, args, cborShapeContext{}, r.bound("ServiceContract"))
	if err != nil {
		return nil, err
	}
	if result.Output == nil {
		return nil, cborRefError("registry_unresolved")
	}
	return result.Output, nil
}

func (r *applicationReference) header(input []byte) (*applicationHeader, error) {
	raw, err := applicationCapture(input, r.bound("ApplicationHeader"))
	if err != nil {
		return nil, err
	}
	v, err := r.wireMap(raw, "ApplicationHeader", cborShapeContext{}, r.bound("ApplicationHeader"))
	if err != nil {
		return nil, err
	}
	code := r.value("ApplicationHeader", v, "message_kind").n
	var kind string
	var variant applicationVariant
	for name, candidate := range r.application.Kinds {
		if candidate.Code == code {
			kind, variant = name, candidate
			break
		}
	}
	if kind == "" {
		return nil, cborRefError("application_kind")
	}
	if len(v.pairs) != len(variant.Fields) {
		return nil, cborRefError("application_fields")
	}
	for _, id := range variant.Fields {
		if cborLookup(v, id) == nil {
			return nil, cborRefError("application_fields")
		}
	}
	for key, expected := range variant.Constants {
		id, err := strconv.ParseUint(key, 10, 64)
		if err != nil {
			return nil, cborRefError("registry_unresolved")
		}
		actual := cborLookup(v, id)
		if actual == nil || actual.major != 0 || actual.n != expected {
			return nil, cborRefError("application_constant")
		}
	}
	if variant.SDKError && r.value("ApplicationHeader", v, "payload_length").n > r.bound(r.application.SDKErrors.PayloadSchema) {
		return nil, cborRefError("application_sdk_error_limit")
	}
	return &applicationHeader{kind: kind, value: v, raw: raw}, nil
}

func (r *applicationReference) response(original, input []byte) (*applicationHeader, error) {
	request, err := r.header(original)
	if err != nil {
		return nil, err
	}
	result, err := r.header(input)
	if err != nil {
		return nil, err
	}
	variant := r.application.Kinds[result.kind]
	if variant.Request != request.kind {
		return nil, cborRefError("application_response_kind")
	}
	for _, name := range []string{"operation_id", "type_id", "request_digest", "service_contract_digest", "control_serial"} {
		if !cborValuesEqual(r.value("ApplicationHeader", request.value, name), r.value("ApplicationHeader", result.value, name)) {
			return nil, cborRefError("application_response_binding")
		}
	}
	limit := r.value("ApplicationHeader", request.value, "response_limit_bytes")
	if limit != nil && !variant.SDKError && r.value("ApplicationHeader", result.value, "payload_length").n > limit.n {
		return nil, cborRefError("application_response_limit")
	}
	return result, nil
}

func (r *applicationReference) contract(header *applicationHeader, requestKind string, input []byte) ([]byte, *cborRefValue, error) {
	raw, err := applicationCapture(input, r.bound("ServiceContract"))
	if err != nil {
		return nil, nil, err
	}
	value, err := r.rawMap("ServiceContract", raw)
	if err != nil {
		return nil, nil, err
	}
	digest, err := r.hash("service_contract_digest", map[string]any{"contract": raw})
	if err != nil {
		return nil, nil, err
	}
	if !bytes.Equal(digest, r.value("ApplicationHeader", header.value, "service_contract_digest").data) || !cborValuesEqual(r.value("ApplicationHeader", header.value, "type_id"), r.value("ServiceContract", value, "type_id")) {
		return nil, nil, cborRefError("application_contract_binding")
	}
	shape, ok := r.application.Policy.ContractVariants[requestKind]
	if !ok {
		return nil, nil, cborRefError("application_contract_variant")
	}
	for key, expected := range shape {
		id, err := strconv.ParseUint(key, 10, 64)
		if err != nil {
			return nil, nil, cborRefError("registry_unresolved")
		}
		actual := cborLookup(value, id)
		if actual == nil || actual.major != 0 || actual.n != expected {
			return nil, nil, cborRefError("application_contract_variant")
		}
	}
	return raw, value, nil
}

func (r *applicationReference) executionDigest(input, contractInput, payloadInput []byte) ([]byte, error) {
	header, err := r.header(input)
	if err != nil {
		return nil, err
	}
	if !slices.Contains(r.application.Policy.ExecutionRequests, header.kind) {
		return nil, cborRefError("application_execution_request")
	}
	contract, definition, err := r.contract(header, header.kind, contractInput)
	if err != nil {
		return nil, err
	}
	_, field, err := r.namedField("ApplicationHeader", "payload_length")
	if err != nil {
		return nil, err
	}
	if field.Max == nil {
		return nil, cborRefError("registry_unresolved")
	}
	payload, err := applicationCapture(payloadInput, uint64(*field.Max))
	if err != nil {
		return nil, err
	}
	if uint64(len(payload)) != r.value("ApplicationHeader", header.value, "payload_length").n {
		return nil, cborRefError("application_payload_length")
	}
	if uint64(len(payload)) > r.value("ServiceContract", definition, "request_max_bytes").n {
		return nil, cborRefError("application_request_limit")
	}
	limit := r.value("ApplicationHeader", header.value, "response_limit_bytes").n
	if limit < r.value("ServiceContract", definition, "min_response_limit_bytes").n || limit > r.value("ServiceContract", definition, "max_response_bytes").n {
		return nil, cborRefError("application_response_limit")
	}
	args := map[string]any{"contract": contract, "payload": payload}
	for _, name := range []string{"message_kind", "operation_id", "type_id", "deadline_at_ms", "admission_mode", "response_limit_bytes"} {
		value := r.value("ApplicationHeader", header.value, name)
		if value.major == 2 {
			args[name] = value.data
		} else {
			args[name] = value.n
		}
	}
	return r.hash("execution_request_digest", args)
}
func (r *applicationReference) verifyExecution(input, contract, payload []byte) error {
	header, err := r.header(input)
	if err != nil {
		return err
	}
	expected, err := r.executionDigest(header.raw, contract, payload)
	if err != nil {
		return err
	}
	if !bytes.Equal(expected, r.value("ApplicationHeader", header.value, "request_digest").data) {
		return cborRefError("application_request_digest")
	}
	return nil
}

func (r *applicationReference) matchErrorSchema(definition, registered []byte) error {
	value, err := r.rawMap("ErrorDefinition", definition)
	if err != nil {
		return err
	}
	domain, ok := r.domains["business_error_schema_digest"]
	if !ok || len(domain.Input.Parts) != 1 || domain.Input.Parts[0].MaxLength == nil {
		return cborRefError("registry_unresolved")
	}
	raw, err := applicationCapture(registered, uint64(*domain.Input.Parts[0].MaxLength))
	if err != nil {
		return err
	}
	digest, err := r.hash(domain.Name, map[string]any{"error_schema": raw})
	if err != nil {
		return err
	}
	if !bytes.Equal(digest, r.value("ErrorDefinition", value, "schema_digest").data) {
		return cborRefError("application_error_schema_mismatch")
	}
	return nil
}
func (r *applicationReference) matchErrorCatalog(input []byte, definitions [][]byte) error {
	contract, err := r.rawMap("ServiceContract", input)
	if err != nil {
		return err
	}
	_, field, err := r.namedField("ServiceContract", "application_error_catalog")
	if err != nil {
		return err
	}
	if field.MaxItems == nil {
		return cborRefError("registry_unresolved")
	}
	if uint64(len(definitions)) > uint64(*field.MaxItems) {
		return cborRefError("application_error_catalog_size")
	}
	catalog := r.value("ServiceContract", contract, "application_error_catalog")
	if len(catalog.items) != len(definitions) {
		return cborRefError("application_error_catalog_mismatch")
	}
	for i, raw := range definitions {
		local, err := r.rawMap("ErrorDefinition", raw)
		if err != nil {
			return err
		}
		if !cborValuesEqual(catalog.items[i], local) {
			return cborRefError("application_error_catalog_mismatch")
		}
	}
	return nil
}

type applicationBusinessResult struct {
	Classification string
	Code           uint64
}

func (r *applicationReference) businessError(original, input, contractInput, payloadInput []byte) (applicationBusinessResult, error) {
	result := applicationBusinessResult{}
	header, err := r.response(original, input)
	if err != nil {
		return result, err
	}
	variant := r.application.Kinds[header.kind]
	if !variant.BusinessError {
		return result, cborRefError("application_error_kind")
	}
	_, contract, err := r.contract(header, variant.Request, contractInput)
	if err != nil {
		return result, err
	}
	payload, err := applicationCapture(payloadInput, r.value("ApplicationHeader", header.value, "payload_length").n)
	if err != nil {
		return result, err
	}
	if uint64(len(payload)) != r.value("ApplicationHeader", header.value, "payload_length").n {
		return result, cborRefError("application_payload_length")
	}
	result.Code = r.value("ApplicationHeader", header.value, "application_error_code").n
	result.Classification = "unknown_application_error"
	for _, entry := range r.value("ServiceContract", contract, "application_error_catalog").items {
		if r.value("ErrorDefinition", entry, "code").n != result.Code {
			continue
		}
		result.Classification = "known_application_error"
		if uint64(len(payload)) > r.value("ErrorDefinition", entry, "max_payload_bytes").n {
			result.Classification = "application_result_decode_failed"
		}
		break
	}
	// No application decoder, payload retention, execution or retry fact.
	return result, nil
}

type applicationSDKResult struct{ Code string }

func (r *applicationReference) sdkError(original, input, payloadInput []byte) (applicationSDKResult, error) {
	result := applicationSDKResult{}
	header, err := r.response(original, input)
	if err != nil {
		return result, err
	}
	if !r.application.Kinds[header.kind].SDKError {
		return result, cborRefError("application_sdk_error_kind")
	}
	name := r.application.SDKErrors.PayloadSchema
	payload, err := applicationCapture(payloadInput, r.bound(name))
	if err != nil {
		return result, err
	}
	if uint64(len(payload)) != r.value("ApplicationHeader", header.value, "payload_length").n {
		return result, cborRefError("application_payload_length")
	}
	value, err := r.rawMap(name, payload)
	if err != nil {
		return result, err
	}
	for code, n := range r.application.SDKCodes {
		if n == r.value(name, value, "code").n {
			stream := header.kind == "execution_stream_sdk_error" || header.kind == "transient_stream_sdk_error"
			if stream && (code == "request_message_aborted" || code == "response_output_stopped") {
				return result, cborRefError("streaming_sdk_error_code")
			}
			if !stream && code == "source_overflow" {
				return result, cborRefError("application_sdk_error_code")
			}
			result.Code = code
			return result, nil
		}
	}
	return result, cborRefError("application_sdk_error_code")
}

File: flowersec-ts/src/v4/applicationReference.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { transportV4Registry } from "../generated/transportV4Registry.js";
import { Application } from "./testSupport/application.js";
import { encode, uint } from "./testSupport/cbor.js";
import type { Value } from "./testSupport/cbor.js";
import { emptyContext } from "./testSupport/shape.js";
import { hex } from "./testSupport/rules.js";
import { root, V4Failure } from "./testSupport/unicode.js";

type Vector = { id: string; hex: string; kind?: string; accept?: boolean; expected_error?: string };
type Corpus = { schema_sha256: string; schema_revision: string; vectors: Vector[] };
const load = (path: string): Corpus => JSON.parse(readFileSync(new URL(`testdata/transport_v4/${path}.json`, root), "utf8")) as Corpus;
const corpus = load("application_headers"), maps = load("corpus"), r = new Application();
const bstr = (value: Uint8Array): Value => ({ kind: "bytes", value });
const field = (name: string, value: Value, key: string): Value => r.requiredValue(name, value, key);
const read = (name: string, input: Uint8Array): Value => r.wireMap(input, name, emptyContext(), 8192n);
const seed = (id: string): Uint8Array => hex(maps.vectors.find(v => v.id === id)!.hex);
const maximum = (kind: string): Value => r.header(hex(corpus.vectors.find(v => v.kind === kind && v.accept)!.hex)).value;
function change(name: string, value: Value, key: string, replacement?: Value): Value {
  if (value.kind !== "map") throw new Error("map fixture");
  const id = r.namedField(name, key)[0], fields = value.value.filter(([key]) => key.kind !== "uint" || key.value !== id);
  if (replacement) fields.push([uint(id), replacement]);
  fields.sort(([a], [b]) => a.kind === "uint" && b.kind === "uint" ? a.value < b.value ? -1 : a.value > b.value ? 1 : 0 : 0);
  return { kind: "map", value: fields };
}
const hset = (value: Value, key: string, replacement?: Value): Value => change("ApplicationHeader", value, key, replacement);
function fail(operation: () => unknown, code?: string): void {
  let caught: unknown; try { operation(); } catch (error) { caught = error; }
  expect(caught).toBeInstanceOf(V4Failure); if (code) expect((caught as V4Failure).code).toBe(code);
}
function digest(name: string, args: Record<string, unknown>): Uint8Array {
  return r.evaluate(name, args, emptyContext(), 8192n).output!;
}
function pair(requestKind: string, responseKind: string, length = 0n, limit = 0n): { request: Value; response: Value } {
  let request = maximum(requestKind), response = maximum(responseKind);
  if (r.namedValue("ApplicationHeader", request, "response_limit_bytes")) request = hset(request, "response_limit_bytes", uint(limit));
  response = hset(response, "payload_length", uint(length));
  return { request, response };
}
function execution(kind: string, payload = Uint8Array.of(1, 2, 3)): { header: Value; contract: Uint8Array; payload: Uint8Array } {
  const suffix = kind === "execution_notify" ? "notify_execution" : kind === "execution_stream_request" ? "stream_execution" : "unary_execution";
  const contract = seed(`service_${suffix}`), definition = read("ServiceContract", contract);
  let header = maximum(kind);
  for (const [key, value] of [
    ["type_id", field("ServiceContract", definition, "type_id")], ["payload_length", uint(BigInt(payload.length))],
    ["service_contract_digest", bstr(digest("service_contract_digest", { contract }))],
    ["response_limit_bytes", field("ServiceContract", definition, "max_response_bytes")],
  ] as const) header = hset(header, key, value);
  header = hset(header, "request_digest", bstr(r.executionDigest(encode(header), contract, payload)));
  return { header, contract, payload };
}

describe("v4.ts_application", () => {
  // v4.ts_application.corpus
  it("independently consumes every exact header and malformed shared candidate", () => {
    expect(maps.schema_sha256).toBe(transportV4Registry.schemaSHA256);
    expect(corpus.schema_revision).toBe(maps.schema_revision);
    expect(corpus.vectors).toHaveLength(589);
    const kinds = new Set<string>(); let rejected = 0;
    for (const v of corpus.vectors) {
      const bytes = hex(v.hex);
      if (v.accept) {
        const result = r.header(bytes); kinds.add(result.kind);
        expect(result.kind, v.id).toBe(v.kind); expect(encode(result.value), v.id).toEqual(bytes);
        expect(bytes.length, v.id).toBe(r.kinds[result.kind]!.max_encoded_bytes);
      } else { rejected++; fail(() => r.header(bytes), v.expected_error?.startsWith("application_") ? v.expected_error : undefined); }
    }
    expect([...kinds].sort()).toEqual(Object.keys(r.kinds).sort()); expect(rejected).toBe(558);
    fail(() => r.header(encode(hset(maximum("execution_notify"), "response_limit_bytes", uint(1n)))), "application_constant");
  });

  // v4.ts_application.response_binding
  it("uses original request kind, identifiers and limits for every response variant", () => {
    for (const [kind, variant] of Object.entries(r.kinds).filter(([, v]) => v.request)) {
      const p = pair(variant.request!, kind); r.response(encode(p.request), encode(p.response));
      for (const name of ["operation_id", "type_id", "request_digest", "service_contract_digest", "control_serial"]) {
        const value = r.namedValue("ApplicationHeader", p.response, name); if (!value) continue;
        const replacement = value.kind === "uint" ? uint(value.value - 1n) : bstr(new Uint8Array(32).fill(7));
        fail(() => r.response(encode(p.request), encode(hset(p.response, name, replacement))), "application_response_binding");
      }
      for (const name of ["deadline_at_ms", "admission_mode", "response_limit_bytes"]) for (const n of [0n, 1n]) fail(() => r.response(encode(p.request), encode(hset(p.response, name, uint(n)))), "application_fields");
      for (const [other, otherVariant] of Object.entries(r.kinds).filter(([other, v]) => !v.request && other !== variant.request)) {
        let request = maximum(other);
        if (otherVariant.fields.includes(Number(r.namedField("ApplicationHeader", "response_limit_bytes")[0]))) request = hset(request, "response_limit_bytes", uint(0n));
        fail(() => r.response(encode(request), encode(p.response)), "application_response_kind");
      }
    }
  });

  // v4.ts_application.execution_digest
  it("binds full execution bytes, exact contract variant, limits and immutable options", () => {
    for (const kind of r.application.policy.execution_requests) {
      const p = execution(kind), raw = encode(p.header); r.verifyExecution(raw, p.contract, p.payload);
      for (const [key, value] of [["deadline_at_ms", uint(17n)], ["admission_mode", uint(0n)], ["operation_id", bstr(new Uint8Array(32))]] as const) fail(() => r.verifyExecution(encode(hset(p.header, key, value)), p.contract, p.payload), "application_request_digest");
      fail(() => r.verifyExecution(raw, p.contract, Uint8Array.of(1, 2, 4)), "application_request_digest");
      fail(() => r.verifyExecution(raw, p.contract, new Uint8Array()), "application_payload_length");
      fail(() => r.executionDigest(encode(hset(p.header, "type_id", uint(900n))), p.contract, p.payload), "application_contract_binding");
      const contract = change("ServiceContract", read("ServiceContract", p.contract), "request_schema_revision", { kind: "text", value: "changed" }), changed = encode(contract);
      fail(() => r.verifyExecution(raw, changed, p.payload), "application_contract_binding");
      const rebound = hset(p.header, "service_contract_digest", bstr(digest("service_contract_digest", { contract: changed })));
      fail(() => r.verifyExecution(encode(rebound), changed, p.payload), "application_request_digest");
    }
    const p = execution("execution_unary_request");
    fail(() => r.executionDigest(encode(hset(p.header, "message_kind", uint(BigInt(r.kinds.execution_stream_request!.code)))), p.contract, p.payload), "application_contract_variant");
    for (const kind of Object.keys(r.kinds).filter(kind => !(r.application.policy.execution_requests as readonly string[]).includes(kind))) fail(() => r.executionDigest(encode(maximum(kind)), p.contract, p.payload), "application_execution_request");
    const small = encode(change("ServiceContract", read("ServiceContract", p.contract), "request_max_bytes", uint(2n)));
    fail(() => r.executionDigest(encode(hset(p.header, "service_contract_digest", bstr(digest("service_contract_digest", { contract: small })))), small, p.payload), "application_request_limit");
    const limited = encode(change("ServiceContract", change("ServiceContract", read("ServiceContract", p.contract), "max_response_bytes", uint(2n)), "min_response_limit_bytes", uint(1n)));
    const limitedHeader = hset(p.header, "service_contract_digest", bstr(digest("service_contract_digest", { contract: limited })));
    for (const n of [0n, 3n]) fail(() => r.executionDigest(encode(hset(limitedHeader, "response_limit_bytes", uint(n))), limited, p.payload), "application_response_limit");
    for (const length of [0, 1048576]) {
      const boundary = execution("execution_unary_request", new Uint8Array(length));
      r.verifyExecution(encode(boundary.header), boundary.contract, boundary.payload);
    }
  });

  // v4.ts_application.sdk_stop
  it("preserves the fixed small-error reservation and never synthesizes execution facts", () => {
    for (const [kind, variant] of Object.entries(r.kinds).filter(([, v]) => v.sdk_error)) for (const [label, code] of Object.entries(r.application.sdk_error_codes)) {
      const body = encode(r.namedMap(r.application.sdk_errors.payload_schema, [["code", uint(BigInt(code))]]));
      const p = pair(variant.request!, kind, BigInt(body.length)), request = encode(p.request), response = encode(p.response);
      const stream = ["execution_stream_sdk_error", "transient_stream_sdk_error"].includes(kind);
      if (stream && ["request_message_aborted", "response_output_stopped"].includes(label)) {
        fail(() => r.sdkError(request, response, body), "streaming_sdk_error_code"); continue;
      }
      if (!stream && label === "source_overflow") {
        fail(() => r.sdkError(request, response, body), "application_sdk_error_code"); continue;
      }
      const result = r.sdkError(request, response, body); expect(result).toEqual({ code: label }); expect(Object.isFrozen(result)).toBe(true);
      expect(Object.keys(result)).toEqual(["code"]);
      for (const length of [0n, 1n, 255n, 256n]) r.response(request, encode(hset(p.response, "payload_length", uint(length))));
      fail(() => r.response(request, encode(hset(p.response, "payload_length", uint(257n)))), "application_sdk_error_limit");
      fail(() => r.sdkError(request, response, new Uint8Array(257)), "application_input_size");
      fail(() => r.sdkError(request, response, body.subarray(0, -1)), "application_payload_length");
      for (const raw of ["a10000", "a10019ffff", "a0", "a200010100", "a200010001", "a1180001", "a1000100", "a100"]) {
        const invalid = hex(raw); fail(() => r.sdkError(request, encode(hset(p.response, "payload_length", uint(BigInt(invalid.length)))), invalid));
      }
      if (r.namedValue("ApplicationHeader", p.request, "response_limit_bytes")) {
        for (const [other, descriptor] of Object.entries(r.kinds).filter(([, v]) => v.request === variant.request && !v.sdk_error)) {
          const alternate = pair(variant.request!, other, 1n); expect(descriptor.sdk_error).toBeUndefined();
          fail(() => r.response(encode(alternate.request), encode(alternate.response)), "application_response_limit");
        }
      }
    }
    const p = pair("transient_unary_request", "transient_unary_response");
    fail(() => r.sdkError(encode(p.request), encode(p.response), new Uint8Array()), "application_sdk_error_kind");
  });

  // v4.ts_application.business_errors
  it("matches exact local catalog bytes and classifies only within the original business limit", () => {
    const schema = hex("a10001"), definition = r.namedMap("ErrorDefinition", [["code", uint(12345n)], ["schema_revision", { kind: "text", value: "error-1" }], ["max_payload_bytes", uint(2n)], ["schema_digest", bstr(digest("business_error_schema_digest", { error_schema: schema }))]]), definitionBytes = encode(definition);
    r.matchErrorSchema(definitionBytes, schema);
    fail(() => r.matchErrorSchema(definitionBytes, hex("a0")), "application_error_schema_mismatch");
    fail(() => r.matchErrorSchema(definitionBytes, new Uint8Array(8193)), "application_input_size");
    for (const semantics of ["execution", "transient"]) for (const shape of ["unary", "stream"]) {
      const contract = encode(change("ServiceContract", read("ServiceContract", seed(`service_${shape}_${semantics}`)), "application_error_catalog", { kind: "array", value: [definition] }));
      r.matchErrorCatalog(contract, [definitionBytes]);
      fail(() => r.matchErrorCatalog(contract, []), "application_error_catalog_mismatch");
      fail(() => r.matchErrorCatalog(contract, Array<Uint8Array>(65).fill(definitionBytes)), "application_error_catalog_size");
      for (const [key, value] of [["code", uint(12346n)], ["schema_revision", { kind: "text", value: "error-2" }], ["max_payload_bytes", uint(3n)], ["schema_digest", bstr(new Uint8Array(32))]] as const) fail(() => r.matchErrorCatalog(contract, [encode(change("ErrorDefinition", definition, key, value))]), "application_error_catalog_mismatch");
      const p = pair(`${semantics}_${shape}_request`, `${semantics}_${shape}_application_error`, 2n, 4n);
      for (const side of ["request", "response"] as const) {
        p[side] = hset(p[side], "service_contract_digest", bstr(digest("service_contract_digest", { contract })));
        p[side] = hset(p[side], "type_id", field("ServiceContract", read("ServiceContract", contract), "type_id"));
      }
      const classify = (length: number, code: bigint) => r.businessError(encode(p.request), encode(hset(hset(p.response, "payload_length", uint(BigInt(length))), "application_error_code", uint(code))), contract, new Uint8Array(length));
      expect(classify(2, 12345n)).toEqual({ classification: "known_application_error", code: 12345n });
      expect(classify(3, 12345n)).toEqual({ classification: "application_result_decode_failed", code: 12345n });
      expect(classify(3, 0xffffffffn)).toEqual({ classification: "unknown_application_error", code: 0xffffffffn });
      const valid = hset(p.response, "application_error_code", uint(12345n));
      fail(() => r.businessError(encode(p.request), encode(valid), contract, new Uint8Array(1)), "application_payload_length");
      fail(() => r.businessError(encode(p.request), encode(valid), contract, new Uint8Array(3)), "application_input_size");
      fail(() => classify(5, 12345n), "application_response_limit");
      p.request = hset(p.request, "response_limit_bytes", uint(0n));
      expect(Object.isFrozen(classify(0, 12345n))).toBe(true); fail(() => classify(1, 12345n), "application_response_limit");
    }
  });

  // v4.ts_application.ownership
  it("bounds copies and owns outputs without executing input metadata getters", () => {
    const p = execution("execution_unary_request"), header = encode(p.header), before = new Uint8Array(header);
    const parsed = r.header(Buffer.from(header)); parsed.bytes.fill(0); expect(header).toEqual(before);
    const source = Buffer.from(header), decoded = r.header(source); source.fill(0); expect(encode(decoded.value)).toEqual(before);
    fail(() => r.header(new Uint8Array(513)), "application_input_size");
    fail(() => r.executionDigest(header, new Uint8Array(8193), p.payload), "application_input_size");
    fail(() => r.executionDigest(header, p.contract, new Uint8Array(1048577)), "application_input_size");
    let traps = 0;
    const proxy = new Proxy(header, { get() { traps++; throw new Error("input trap"); }, getPrototypeOf() { traps++; throw new Error("input trap"); } });
    const revoked = Proxy.revocable(header, {}); revoked.revoke();
    for (const input of [proxy, revoked.proxy]) fail(() => r.header(input), "application_input_bytes");
    const detached = new Uint8Array(header); structuredClone(detached, { transfer: [detached.buffer] });
    fail(() => r.header(detached), "application_input_bytes");
    for (const key of ["length", "byteLength", "byteOffset", "buffer"]) {
      const input = new Uint8Array(header); Object.defineProperty(input, key, { get() { traps++; throw new Error("metadata trap"); } });
      fail(() => r.header(input), "application_input_bytes");
    }
    const backing = new ArrayBuffer(header.length + 4), input = new Uint8Array(backing, 2, header.length); input.set(header);
    Object.defineProperty(backing, "byteLength", { get() { traps++; throw new Error("backing trap"); } });
    expect(r.header(input).bytes).toEqual(header); expect(traps).toBe(0);
  });

  // v4.ts_application.properties
  // Allow the bounded corpus to finish alongside the other language gates.
  it("keeps mutated header validation deterministic and detached across 4096 cases", () => {
    const positives = corpus.vectors.filter(v => v.accept); let state = 0x235ba919;
    const next = (): number => { state ^= state << 13; state ^= state >>> 17; state ^= state << 5; return state >>> 0; };
    for (let i = 0; i < 4096; i++) {
      const raw = hex(positives[next() % positives.length]!.hex); raw[next() % raw.length]! ^= 1 << (next() % 8);
      const before = new Uint8Array(raw);
      const result = (): string => { try { const decoded = r.header(raw); expect(encode(decoded.value)).toEqual(before); decoded.bytes.fill(0); return decoded.kind; } catch (error) { expect(error).toBeInstanceOf(V4Failure); return `error:${(error as V4Failure).code}`; } };
      expect(result()).toBe(result()); expect(raw).toEqual(before);
    }
  }, 15_000);
});
